pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,251 @@
+package nodelabeller
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+const (
+	labelNamespace    = "feature.node.kubernetes.io"
+	labellerNamespace = "node-labeller"
+)
+
+var nodeLabellerVolumePath = "/var/lib/kubevirt-node-labeller"
+
+//NodeLabeller struct holds informations needed to run node-labeller
+type NodeLabeller struct {
+	lock              sync.Mutex
+	kvmController     *device_manager.DeviceController
+	configMapInformer cache.SharedIndexInformer
+	nodeInformer      cache.SharedIndexInformer
+	clientset         kubecli.KubevirtClient
+	Queue             workqueue.RateLimitingInterface
+	host              string
+	namespace         string
+	logger            *log.FilteredLogger
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, nodeInformer, configMapInformer cache.SharedIndexInformer, clientset kubecli.KubevirtClient, host, namespace string) *NodeLabeller {
+	return &NodeLabeller{
+		kvmController:     kvmController,
+		configMapInformer: configMapInformer,
+		nodeInformer:      nodeInformer,
+		clientset:         clientset,
+		Queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		host:      host,
+		namespace: namespace,
+		logger:    log.DefaultLogger(),
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.Queue.ShutDown()
+	n.logger.Infof("node-labeller is running")
+
+	if !n.kvmController.NodeHasDevice(device_manager.KVMPath) {
+		n.logger.Infof("Node-labeller cannot work without KVM device.")
+		return
+	}
+
+	n.configMapInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			n.Queue.AddRateLimited(obj)
```

Comment:
two things here.

1. You can use n.Queue.Add() instead of AddRateLimited. We only need to rate limit errors.

2. queuing the actual object here is unexpected. It technically compiles, but probably doesn't give the results you're looking for.

take a look at how `pkg/virt-controller/watch/node.go` does the node callbacks. 

```
        node := obj.(*v1.Node)
        key, err := controller.KeyFunc(node) 
        if err != nil {
                logger.Object(node).Reason(err).Error("Failed to extract key from node.")
        }
        c.Queue.Add(key)
```

Then in the execute function the key is used to retrieve the node from cache. 
```
obj, nodeExists, err := c.nodeInformer.GetStore().GetByKey(key)
if nodeExists {
                node = obj.(*v1.Node)
                logger = logger.Object(node)
        } else {
                logger = logger.Key(key, "Node")
        }
```